priv/templates/phx.gen.auth/scope.ex

```diff
@@ -0,0 +1,31 @@
+defmodule <%= inspect context.module %>.<%= inspect schema.alias %>Scope do
+  @moduledoc """
+  Defines the scope of the caller to be used throughout the app.
+
+  The `<%= inspect context.module %>.<%= inspect schema.alias %>Scope` allows public interfaces to receive
+  information about the caller, such as if the call is initiated from an
+  end-user, and if so, which user. Additionally, such a scope can carry fields
+  such as "super user" or other privileges for use as authorization, or to
+  ensure specific code paths can only be access for a given scope.
+
+  It is useful for logging as well as for scoping pubsub subscriptions and
+  broadcasts when a caller subscribes to an interface or performs a particular
+  action.
+
+  Feel free to extend the fields on this struct to fit the needs of
+  growing application requirements.
+  """
+
+  alias <%= inspect schema.module %>
+
+  defstruct user: nil
+
+  @doc """
+  Creates a scope for the given <%= schema.singular %>.
+  """
+  def for_<%= schema.singular %>(nil), do: %__MODULE__{user: nil}
```

Comment:
Honestly, I would prefer to not have this pattern. If we are thinking about the type system, it means that now we need to handle about the possibility of user being nil everywhere. Or we have a scope with user, or we have no scope.

Reply:
I guess it depends if we think that it is common to want scoped functionality for unauthenticated requests. I think it would be common to have one scope struct with fields that can either be nil or not instead of different scope structs.

The alternative is to have a `for_user(nil), do: nil` or the need to change all UserAuth functions to something like

```elixir
defp mount_current_scope(socket, session) do
    Phoenix.Component.assign_new(socket, :current_scope, fn ->
      with user_token when is_binary(user_token) <- session["user_token"],
           user when not is_nil(user) <- Accounts.get_user_by_session_token(user_token) do
          AuthScope.for_user(user)
       end
    end)
  end
```

and only call the scope function if we have a user in the first place.

Reply:
> I think it would be common to have one scope struct with fields that can either be nil or not instead of different scope structs.

I agree, but I would prefer to not assume and start with the most streamlined version. For example, in teams, I believe we always have org and user.